src/core/expression/qgsexpressionfunction.cpp

```diff
@@ -4920,13 +4916,46 @@ static QVariant executeGeomOverlay( const QVariantList &values, const QgsExpress
     return QVariant();
   }
 
+  QgsFeatureRequest request; // TODO only required attributes
+  QString filterString;
+  node = QgsExpressionUtils::getNode( values.at( 2 ), parent );
+  ENSURE_NO_EVAL_ERROR
+  filterString = node->dump();
+  if ( filterString != "NULL" )
+  {
+    request.setFilterExpression( filterString ); //filter cached features
+  }
+
+  node = QgsExpressionUtils::getNode( values.at( 3 ), parent ); //in expressions overlay functions throw the exception: Eval Error: Cannot convert '' to int
+  ENSURE_NO_EVAL_ERROR
+  QString limitString = node->dump();
```

Comment:
I think you need to go for the following:

```suggestion
  QVariant limitValue = node->eval( parent, context );
  ENSURE_NO_EVAL_ERROR
  qlonglong limit = QgsExpressionUtils::getIntValue( limitValue, parent );
```

The issue is, that we need to do "lazy evaluation" in this method. We can't let the expression evaluate everything because we need to take care of the "filter" string manually and evaluate it with a different context.